flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java

```diff
@@ -0,0 +1,342 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.api.functions.sink;
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.api.common.state.ListState;
+import org.apache.flink.api.common.state.ListStateDescriptor;
+import org.apache.flink.api.common.typeinfo.TypeHint;
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.runtime.state.CheckpointListener;
+import org.apache.flink.runtime.state.FunctionInitializationContext;
+import org.apache.flink.runtime.state.FunctionSnapshotContext;
+import org.apache.flink.streaming.api.checkpoint.CheckpointedFunction;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This is a recommended base class for all of the {@link SinkFunction} that intend to implement exactly-once semantic.
+ * It does that by implementing two phase commit algorithm on top of the {@link CheckpointedFunction} and
+ * {@link CheckpointListener}. User should provide custom TXN (transaction handle) and implement abstract methods
+ * handling this transaction handle.
+ *
+ * @param <IN> Input type for {@link SinkFunction}
+ * @param <TXN> Transaction to store all of the information required to handle a transaction (must be Serializable)
+ */
+@PublicEvolving
+public abstract class TwoPhaseCommitSinkFunction<IN, TXN extends Serializable>
+		extends RichSinkFunction<IN>
+		implements CheckpointedFunction, CheckpointListener {
+
+	private static final Logger LOG = LoggerFactory.getLogger(TwoPhaseCommitSinkFunction.class);
+
+	protected final ListStateDescriptor<TransactionAndCheckpoint<TXN>> pendingCommitTransactionsDescriptor;
+	protected final ListStateDescriptor<TXN> pendingTransactionsDescriptor;
+
+	protected final List<TransactionAndCheckpoint<TXN>> pendingCommitTransactions = new ArrayList<>();
+
+	@Nullable
+	protected TXN currentTransaction;
+	protected ListState<TXN> pendingTransactionsState;
+	protected ListState<TransactionAndCheckpoint<TXN>> pendingCommitTransactionsState;
+
+	public TwoPhaseCommitSinkFunction(Class<TXN> txnClass) {
```

Comment:
I'm not too sure we actually want to expose this specific constructor.
I'm not really fond of the fact that the user leaves the decision of how to serialize the state completely to this base class. One disadvantage I can think immediately of that is that the `TransactionAndCheckpoint` is not a POJO, and therefore Flink will try to serialize it using Kryo into state. In general, it is a bad idea to use Kryo for these kinds of persisted data.